clang/lib/AST/Expr.cpp

```diff
@@ -648,6 +649,85 @@ UniqueStableNameDiscriminator(ASTContext &, const NamedDecl *ND) {
   return llvm::None;
 }
 
+/// Compute a unique name that is consumable by sycl_vxx
+std::string computeUniqueSYCLVXXName(StringRef Demangle, bool shouldRename) {
+  if (!shouldRename)
+    return Demangle.str();
+
+  /// VXX has a maximum of 64 character for the name of the kernel function
+  /// plus the name of one parameter.
+  /// Those characters need to be used wisely to prevent name collisions.
+  /// It is also useful to use a name that is understandable by the user,
+  /// so we add only 8 character of hash and only if needed.
+  /// The first character cannot be an underscore or a digit.
+  /// An underscore can only be found between two non underscore character
+  /// (No __ or undersocre at end or start of name)
+  constexpr unsigned MaxVXXSize = 30;
+  /// Some transformations might make 2 kernel identifiers the same.
+  /// Allow adding a hash when such transformations are made to avoid possible
+  /// name conflict.
+  bool ForceHash = false;
+
+  std::string Result;
+  Result.reserve(Demangle.size());
+
+  for (char c : Demangle) {
+    if (!isAlphanumeric(c)) {
+      // Do not repeat _ in the cleaned-up name
+      if (!Result.empty() && Result.back() == '_')
+        continue;
+      c = '_';
+    }
+    Result.push_back(c);
+  }
+
+  // Replace first kernel character name by a 'k' to be compatible with SPIR
```

Comment:
Is this the right reason? Does SPIR required `k` and not `a`?

Reply:
it fail if the identifier starts with an `_` or a number. replacing it with a `k` is an arbitrary decision, it could have been replaced with an `a`